luarules/mission_api/types.lua

```diff
@@ -0,0 +1,711 @@
+
+    return out
+end
+
+----------------------------------------------------------------
+
+--- Starts the timer. If paused, will continue at the state the timer was paused
+function Timer:start()
+    self.running = true
+end
+
+----------------------------------------------------------------
+
+--- Pauses the timer, does not reset the timer's current state
+function Timer:pause()
+    self.running = false
+end
+
+----------------------------------------------------------------
+
+--- Stops the timer and resets its current state
+function Timer:stop()
+    self.count = 0
+    self.running = false
+end
+
+----------------------------------------------------------------
+
+--- Polls the timer, incrementing its count and calling the appropriate functions
+function Timer:poll()
+    if not self.running then return end
+
+    self.count = self.count + 1
+
+    if self.onUpdate then self.onUpdate(self.length / self.count) end
+
+    if self.count == self.length then
+        if self.onFinished then self.onFinished() end
+        if not self.loop then self.running = false end
+        self.count = 0
+    end
+end
+
+--============================================================--
+
+-- Unit
+
+--============================================================--
+
+--- Metatable for Unit objects
+local Unit = {
+    __name = 'Unit', -- Meta name for type checking
+
+    TYPE = {
+        NAME = 0, -- All units tracked with the ID as the tracked name
+        ID = 1, -- Only the unit with the ID as UnitID
+        DEF_ID = 2, -- All units with the ID as UnitDefID
+        DEF_NAME = 3, -- All units wit hthe ID as UnitDefName
+    },
+
+    type = 0, -- The type of ID given, must be Unit.TYPE
+    ID = 0, -- The name or ID of the given type
+    team = Spring.ALL_UNITS, -- The teamID to use for filtering units
+}
+Unit.__index = Unit
+
+----------------------------------------------------------------
+
+--- Creates a new instance of Unit
+-- @tparam Unit.TYPE type The type of ID given
+-- @tparam string|number ID The name or ID of the given type
+-- @number team The teamID to use for filtering units
+-- @treturn table Instance derived from the Unit metatable
+function Unit:new(type, ID, team)
+    local object = {}
+    setmetatable(object, self)
+
+    object.type = type or self.type
+    object.ID = ID or self.ID
+    object.team = team or self.team
+
+    return object
+end
+
+----------------------------------------------------------------
+
+--- Validates the Unit's parameters
+-- @string file The name of the file to be used for logging validity errors
+-- @string module The name of the module to be used for logging validity errors
+-- @treturn boolean False if any validity errors are found
+function Unit:validate(file, module)
+    file = file or 'types.lua'
+    module = module or 'Mission API'
+
+    if self.ID == nil then
+        ErrorMissingField(file, module, self.__name, 'ID')
+        return false
+    end
+
+    if not self.type then
```

Comment:
This is going to fail for self.type == Unit.TYPE.NAME. Not sure why I 0-indexed the TYPE enum and then didn't use 'if self.type == nil'. It either needs to be nil.checked or the TYPE enums need to be 1-indexed